Fix cannot pickle module object error in PythonVirtualenvOperator

ephraimbuddy opened the pull request:
When the system_site_packages arg is set and the python_callable has wildcard keyword parameters, the task fails with error that it can't
pickle module object. Here we warn users not to use both

Closes: https://github.com/apache/airflow/issues/20974

cc: @uranusjr WDYT


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

ashb commented on tests/operators/test_python.py:
```diff
@@ -999,6 +999,21 @@ def test_lambda(self):
             PythonVirtualenvOperator(python_callable=lambda x: 4, task_id='task', dag=self.dag)
 
+    def test_system_site_packages_with_wildcard_callable(self):
+        def mycallable(**kwargs):
```
This is...... surprising. Like why would `**kwargs` and site packages cause this behaviour?!

ephraimbuddy replied:
Spent time on it trying to figure out why but it seems that with site packages, the many airflow pickleable objects are no longer pickleable. If you remove everything from https://github.com/apache/airflow/blob/2d6282d6b7d8c7603e96f0f28ebe0180855687f3/airflow/operators/python.py#L358 except config and params, it works with site packages

ashb replied:
Ohhhhh, it's cos of what we pass in those kwargs. That makes sense.

ephraimbuddy commented on tests/operators/test_python.py:
```diff
@@ -1053,7 +1068,7 @@ def f(
             pass
 
-        self._run_as_operator(f, use_dill=True, system_site_packages=True, requirements=None)
+        self._run_as_operator(f, use_dill=True, system_site_packages=False, requirements=None)
```
I'm surprised that this test passes before

ephraimbuddy committed:
Fix cannot pickle module object error in PythonVirtualenvOperator

When the system_site_packages arg is set and the python_callable has wildcard keyword parameters, the task fails with error that it can't
pickle module object

uranusjr commented on airflow/operators/python.py:
```diff
@@ -517,6 +521,16 @@ def __deepcopy__(self, memo):
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        self.pickling_library = dill if self.use_dill else pickle
```
Hmm this attribute should probably have been an `@property` to begin with. In which case we don’t even need to worry about pickling it at all.
